Python 3.10 compatibility broken

## Issue

Conda-forge build:

```
_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/python3.10/site-packages/impact/model/base.py:4: in <module>
    from typing import Any, Generic, TypeVar, Self
E   ImportError: cannot import name 'Self' from 'typing' ($PREFIX/lib/python3.10/typing.py)
```

This should be using `typing_extensions`.

## Extra

But we test on 3.10 and up, right?

Well, no, that mechanism was broken since the switch to pyproject, it appears.
The `environment.yaml` should have had a minimum pin for Python, which gets automatically replaced in the test suite with the appropriate Python version (to avoid building the env twice, etc.):
https://github.com/ChristopherMayes/lume-impact/blob/3cc548de89d5cb2dddd41bffdf705a730e81aa58/.github/actions/conda-setup/action.yml?plain=1#L36-L43

Unfortunately, the `environment.yaml` instead has a fixed pin which breaks that:

https://github.com/ChristopherMayes/lume-impact/blob/3cc548de89d5cb2dddd41bffdf705a730e81aa58/environment.yml?plain=1#L6

So:

1. Switch to typing_extensions
2. Fix the environment yaml to be a min pin

--- a/.github/env-impact-only.yml
+++ b/.github/env-impact-only.yml
@@ -3,6 +3,9 @@ name: lume-impact-dev
 channels:
   - conda-forge
 dependencies:
+  # python gets pulled in via bmad -> xraylib -> numpy anyway,
+  # set this up so CI can control it at least
+  - python>=3.10
   - impact-t=*=mpi_openmpi*
   - impact-z=*=mpi_openmpi*
   # Well, also bmad for PyTao tests

--- a/.github/env-impact-only.yml
+++ b/.github/env-impact-only.yml
@@ -3,7 +3,6 @@ name: lume-impact-dev
 channels:
   - conda-forge
 dependencies:
-  - pip
   - impact-t=*=mpi_openmpi*
   - impact-z=*=mpi_openmpi*
   # Well, also bmad for PyTao tests

--- a/.github/workflows/tests-pip.yml
+++ b/.github/workflows/tests-pip.yml
@@ -47,5 +47,5 @@ jobs:
         shell: bash -l -e -o pipefail {0}
         run: |
           echo -e '## Test results\n\n```' >> "$GITHUB_STEP_SUMMARY"
-          pytest -v 2>&1 | tee -a "$GITHUB_STEP_SUMMARY"
+          python -m pytest -v 2>&1 | tee -a "$GITHUB_STEP_SUMMARY"
           echo '```' >> "$GITHUB_STEP_SUMMARY"

--- a/.github/workflows/tests-pip.yml
+++ b/.github/workflows/tests-pip.yml
@@ -21,7 +21,7 @@ jobs:
           fetch-depth: 0
       - uses: ./.github/actions/conda-setup
         with:
-          python-version: ""
+          python-version: ${{ matrix.python-version }}
           filename: .github/env-impact-only.yml
           pip-package: "" # don't install lume-impact here
       - uses: actions/setup-python@v6

--- a/environment.yml
+++ b/environment.yml
@@ -3,7 +3,7 @@ name: lume-impact-dev
 channels:
   - conda-forge
 dependencies:
-  - python=3.12
+  - python>=3.10
   - bokeh
   - distgen
   - impact-t=*=mpi_openmpi*

--- a/impact/model/base.py
+++ b/impact/model/base.py
@@ -1,7 +1,8 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
-from typing import Any, Generic, TypeVar, Self
+from typing import Any, Generic, TypeVar
+from typing_extensions import Self
 
 from pydantic import BaseModel, model_validator
 

